tests/losses/test_supcon_loss.py

```diff
@@ -0,0 +1,133 @@
+            embedding_angles = [0, 10, 20, 50, 60, 80]
+            embeddings = torch.tensor(
+                [c_f.angle_to_coord(a) for a in embedding_angles],
+                requires_grad=True,
+                dtype=dtype,
+            ).to(
+                TEST_DEVICE
+            )  # 2D embeddings
+
+            labels = torch.LongTensor([0, 0, 0, 1, 1, 2])
+
+            lossA = loss_funcA(embeddings, labels)
+            lossB = loss_funcB(embeddings, labels)
+
+            anchor_pos_pairs = {
+                0: [(0, 1), (0, 2)],
+                1: [(1, 0), (1, 2)],
+                2: [(2, 0), (2, 1)],
+                3: [(3, 4)],
+                4: [(4, 3)],
+            }
+            neg_pairs = [
+                (0, 3),
+                (0, 4),
+                (0, 5),
+                (1, 3),
+                (1, 4),
+                (1, 5),
+                (2, 3),
+                (2, 4),
+                (2, 5),
+                (3, 0),
+                (3, 1),
+                (3, 2),
+                (3, 5),
+                (4, 0),
+                (4, 1),
+                (4, 2),
+                (4, 5),
+                (5, 0),
+                (5, 1),
+                (5, 2),
+                (5, 3),
+                (5, 4),
+            ]
+
+            total_lossA = 0
+            total_lossB = 0
+            for a1, pos_pairs in anchor_pos_pairs.items():
+                anchor_lossA = 0
+                anchor_lossB = 0
+                for __, p in pos_pairs:
+                    anchor, positive = embeddings[a1], embeddings[p]
+                    numeratorA = torch.exp(torch.matmul(anchor, positive) / temperature)
+                    numeratorB = torch.exp(
+                        -torch.sqrt(torch.sum((anchor - positive) ** 2)) / temperature
+                    )
+                    denominatorA = 0
+                    denominatorB = 0
+                    for a2, n in pos_pairs + neg_pairs:
+                        if a2 == a1:
+                            negative = embeddings[n]
+                        else:
+                            continue
+                        denominatorA += torch.exp(torch.matmul(anchor, negative) / temperature)
+                        denominatorB += torch.exp(
+                            -torch.sqrt(torch.sum((anchor - negative) ** 2)) / temperature
+                        )
+                    anchor_lossA += torch.log(numeratorA / denominatorA)
+                    anchor_lossB += torch.log(numeratorB / denominatorB)
+                total_lossA += -anchor_lossA / len(pos_pairs)
+                total_lossB += -anchor_lossB / len(pos_pairs)
+            total_lossA /= len(anchor_pos_pairs)
+            total_lossB /= len(anchor_pos_pairs)
+            rtol = 1e-2 if dtype == torch.float16 else 1e-5
+            self.assertTrue(torch.isclose(lossA, total_lossA, rtol=rtol))
+            self.assertTrue(torch.isclose(lossB, total_lossB, rtol=rtol))
```

Comment:
I already had implemented this before I saw your recent changes on `tests/losses/test_ntxent_loss.py`. Should we change this to match it?